src/utilcode/longfilepathwrappers.cpp

```diff
@@ -0,0 +1,1307 @@
+                hModule,
+                buffer.OpenUnicodeBuffer(size - 1),
+                (DWORD)size
+                );
+          
+        }
+        
+
+        lastError = GetLastError();
+        buffer.CloseBuffer(ret);
+    }
+    EX_CATCH_HRESULT(hr);
+
+    if (hr != S_OK)
+    {
+        SetLastError(hr);
+    }
+    else if (ret == 0)
+    {
+        SetLastError(lastError);
+    }
+
+    return ret;
+    
+
+}
+
+UINT WINAPI GetTempFileNameWrapper(
+    _In_  LPCTSTR lpPathName,
+    _In_  LPCTSTR lpPrefixString,
+    _In_  UINT    uUnique,
+    SString&  lpTempFileName
+    )
+{
+    HRESULT hr = S_OK;
+    UINT ret = 0;
+    DWORD lastError;
+    EX_TRY
+    {
+        //Change the behaviour in Redstone to retry
+        COUNT_T size = MAX_LONGPATH;
+        
+        ret  = GetTempFileNameW(
+            lpPathName,
+            lpPrefixString,
+            uUnique,
+            lpTempFileName.OpenUnicodeBuffer(size - 1)
+            );
+
+        lastError = GetLastError();
+        lpTempFileName.CloseBuffer(ret);
+    }
+    EX_CATCH_HRESULT(hr);
+
+    if (hr != S_OK)
+    {
+        SetLastError(hr);
+    }
+    else if (ret == 0)
+    {
+        SetLastError(lastError);
+    }
+
+    return ret;
+}
+DWORD WINAPI GetTempPathWrapper(
+    SString& lpBuffer
+    )
+{
+    HRESULT hr = S_OK;
+    DWORD ret = 0;
+    DWORD lastError;
+    EX_TRY
+    {
+        //Change the behaviour in Redstone to retry
+        COUNT_T size = MAX_LONGPATH;
+
+        ret = GetTempPathW(
+            size,
+            lpBuffer.OpenUnicodeBuffer(size - 1)
+            );
+
+        lastError = GetLastError();
+        lpBuffer.CloseBuffer(ret);
+    }
+    EX_CATCH_HRESULT(hr);
+
+    if (hr != S_OK)
+    {
+        SetLastError(hr);
+    }
+    else if (ret == 0)
+    {
+        SetLastError(lastError);
+    }
+
+    return ret;
+}
+
+DWORD WINAPI GetCurrentDirectoryWrapper(
+    SString&  lpBuffer
+    )
+{
+    HRESULT hr = S_OK;
+    DWORD ret = 0;
+    DWORD lastError;
+    EX_TRY
+    {
+        //Change the behaviour in Redstone to retry
+        COUNT_T size = MAX_LONGPATH;
+
+        ret = GetCurrentDirectoryW(
+            size, 
+            lpBuffer.OpenUnicodeBuffer(size - 1)
+            );
+
+        lastError = GetLastError();
+        lpBuffer.CloseBuffer(ret);
+    }
+    EX_CATCH_HRESULT(hr);
+
+    if (hr != S_OK)
+    {
+        SetLastError(hr);
+    }
+    else if (ret == 0)
+    {
+        SetLastError(lastError);
+    }
+
+    return ret;
+}
+
+DWORD WINAPI GetEnvironmentVariableWrapper(
+    _In_opt_  LPCTSTR lpName,
+    _Out_opt_ SString&  lpBuffer
+    )
+{
+    HRESULT hr = S_OK;
+    DWORD ret = 0;
+    DWORD lastError;
+    EX_TRY
+    {
+        
+        COUNT_T size = lpBuffer.GetUnicodeAllocation() + 1;
+
+        ret = GetEnvironmentVariableW(
+            lpName, 
+            lpBuffer.OpenUnicodeBuffer(size - 1),
+            size
+            );
+
+        if (ret > size)
```

Comment:
Replace if with while, since the environment variable can be changed by another thread between two calls to GetEnvironmentVariableW.